Add @cover annotation in test-load file and add support for codecov

sarthak-19 opened the pull request:
## Summary
This PR adds `@cover` annotation for test cases and add .yml file that can be used to configure codecov for automated test coverage in github actions.
<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Addresses https://github.com/WordPress/performance/issues/1284

## Relevant technical choices

<!-- Please describe your changes. -->

I've addressed the following file `plugins/webp-uploads/tests/test-load.php`  for adding `@cover` annotation, if the approach is approved will extend this to other files.

<!--
For maintainers only, please make sure:

- PR has a `[Type]` label.
- PR has a plugin-specific milestone, or the `no milestone` label if it does not apply to any specific plugin.
- PR has a changelog-friendly title, or the `skip changelog` label if it should not be mentioned in the plugin's changelog.
-->

Commit:
Add @cover annotation in test-load file and add support for codecov for test coverage

Commit:
Addressed merge conflict

swissspidy commented on composer.json:
```diff
@@ -31,8 +31,10 @@
-    "phpstan/phpstan-strict-rules": "^1.6"
+    "phpstan/php-8-stubs": "^0.4.0",
+    "phpstan/phpstan-strict-rules": "^1.6",
+    "phpunit/phpunit": "^9.5",
```
This isn't needed, PHPUnit is installed via `yoast/phpunit-polyfills` above

swissspidy commented on composer.json:
```diff
@@ -31,8 +31,10 @@
+    "phpstan/php-8-stubs": "^0.4.0",
+    "phpstan/phpstan-strict-rules": "^1.6",
+    "phpunit/phpunit": "^9.5",
+    "codecov/codecov": "^1.0"
```
This isn't needed either, nor does it actually exist

swissspidy commented on .github/workflows/codecov-coverage.yml:
```diff
@@ -0,0 +1,39 @@
+name: PHP Test and Coverage
```
Can we run coverage in the existing `php-test-plugins.yml` workflow instead?

swissspidy commented:
@sarthak-19 Welcome, and thanks for opening this! I just stumbled upon the PR and left a couple of comments for things that caught my eye.

Commit:
Add coverage in the existing php-test-plugins.yml and removed redundant code from composer.json

Commit:
Removed redundant code from composer.json

sarthak-19 commented:
> @sarthak-19 Welcome, and thanks for opening this! I just stumbled upon the PR and left a couple of comments for things that caught my eye.

@swissspidy Thank you so much for the feedback. I've addressed those and looking forward to keep improving 😄

Commit:
Merge branch 'trunk' into fix/add_cover_annotation_test

westonruter commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +68,20 @@ jobs:
         run: npm run wp-env run tests-cli -- --env-cwd="wp-content/plugins/$(basename $(pwd))" composer install --no-interaction --no-progress
       - name: Running single site unit tests
-        run: npm run test-php
+        run: npm run test-php -- --coverage-clover=coverage.xml
```
Since collecting code coverage takes much longer, as I recall, should the coverage only happen for one of the jobs instead of for all of them?

sarthak-19 replied:
Yes you're right, collecting code coverage takes time, but I'm not quite sure that for which job it'll be ideal to run.
Should I run for `PHP v8.0` and if this is a viable option, then will the below implementation work since we are using the latest WP version anyways?
```yml

      - name: Running single site unit tests with coverage
        if: matrix.php == '8.0'
        run: npm run test-php -- --coverage-clover=coverage.xml
      - name: Running multisite unit tests with coverage
        if: matrix.php == '8.0'
        run: npm run test-php-multisite -- --coverage-clover=coverage-multisite.xml

      - name: Upload single site coverage reports to Codecov
        if: matrix.php == '8.0'
        uses: codecov/codecov-action@v4
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: coverage.xml
          flags: unittests
          name: ${{ matrix.php }}-single-site-coverage
      - name: Upload multisite coverage reports to Codecov
        if: matrix.php == '8.0'
        uses: codecov/codecov-action@v2
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: coverage-multisite.xml
          flags: multisite
          name: ${{ matrix.php }}-multisite-coverage
```

cc : @westonruter

westonruter replied:
You may want to refer to what was done for the AMP plugin: https://github.com/ampproject/amp-wp/blob/60a655cdad9e3431b47b2dcc3b2e5ddb86ac5c5e/.github/workflows/build-test-measure.yml#L261-L337

Note the new `coverage` matrix variable which is only set to `true` for one configuration. This would be better than having to test against a specific PHP version everywhere. Yes, I think running coverage on the most recent PHP version would be best as it should complete the fastest. So maybe do it in PHP 8.2?

cc @thelovekesh

pbearne commented:
I am really glad to see I have been try to get into core ten+ years

Commit:
Update workflow to run code coverage using matrix strategy

sarthak-19 committed:
Merge branch 'trunk' into fix/add_cover_annotation_test

sarthak-19 commented:
> Note the new coverage matrix variable which is only set to true for one configuration. This would be better than having to test against a specific PHP version everywhere. Yes, I think running coverage on the most recent PHP version would be best as it should complete the fastest. So maybe do it in PHP 8.2?

Apologies for the delayed response, made the suggested changes.
Does this works or any other changes are to be made, highly appreciate your insights here.

cc : @westonruter

westonruter commented:
@sarthak-19 That looks right!

westonruter committed:
Add EOF line break to codecov.yml

westonruter commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          files: coverage-multisite.xml
+          flags: multisite
```
Is `multisite` a valid flag?  Shouldn't this be `unittests`?

```suggestion
          flags: unittests
```

felixarntz replied:
These are arbitrary values to separate test scenarios, so I think passing `multisite` makes sense here (separating from "single site").

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -43,11 +43,15 @@ jobs:
             wp: '6.5'
           - php: '8.3'
             wp: 'trunk'
+          - php: '8.2'
+            wp: 'latest'
+            coverage: true
```
I wonder, given we include this as an extra scenario, should we maybe remove `8.2` from the regular `php` list above? Otherwise we'd be running the exact same test (WP latest, PHP 8.2) twice, only with and without code coverage, which shouldn't change the outcome of the tests.

If there's any benefit to keeping both (which would only be because of the implications that running code coverage could have), then I think the better choice would be to implement a separate workflow entirely.

sarthak-19 replied:
Yes make sense, removing the 8.2 version from PHP list.

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
-        run: npm run test-php
+        run: |
+          if [ "${{ matrix.coverage }}" == "true" ]; then
+            npm run test-php -- --coverage-clover=coverage.xml
```
Maybe append commit SHA to avoid chance of conflicts?
```suggestion
            npm run test-php -- --coverage-clover=coverage-${{ github.sha }}.xml
```

westonruter replied:
How would there be a conflict? Each GHA job should run from a pristine state, no? For the `coverage.xml` file to persist across runs this would require some explicit caching.

felixarntz replied:
Probably. I find this safer, but either works for me.

sarthak-19 replied:
> I find this safer.

I think better safe than sorry.
So implementing this as well. 😀

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
-        run: npm run test-php-multisite
+        run: |
+          if [ "${{ matrix.coverage }}" == "true" ]; then
+            npm run test-php-multisite -- --coverage-clover=coverage-multisite.xml
```
See above:
```suggestion
            npm run test-php-multisite -- --coverage-clover=coverage-multisite-${{ github.sha }}.xml
```

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
+        uses: codecov/codecov-action@v4
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          files: coverage.xml
```
See above:
```suggestion
          files: coverage-${{ github.sha }}.xml
```

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
+        uses: codecov/codecov-action@v4
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          files: coverage-multisite.xml
```
See above:
```suggestion
          files: coverage-multisite-${{ github.sha }}.xml
```

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          files: coverage.xml
+          flags: unittests
```
How about we use `single` instead of `unittests`, since the differentiator between the two datasets is "single site" vs "multisite", not "unit tests" vs "multisite".
```suggestion
          flags: single
```

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          files: coverage.xml
+          flags: unittests
+          name: ${{ matrix.php }}-single-site-coverage
```
Let's make sure we're alerted of failures for coverage reporting:
```suggestion
          name: ${{ matrix.php }}-single-site-coverage
          fail_ci_if_error: true
```

felixarntz commented on .github/workflows/php-test-plugins.yml:
```diff
@@ -68,6 +72,32 @@ jobs:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          files: coverage-multisite.xml
+          flags: multisite
+          name: ${{ matrix.php }}-multisite-coverage
```
See above:
```suggestion
          name: ${{ matrix.php }}-multisite-coverage
          fail_ci_if_error: true
```

felixarntz reviewed (commented):
@sarthak-19 Thank you for the PR! This looks quite solid, a few suggestions and questions.

Commit:
Addressed feedback :
- Implemented alerted of failures.
- append commit SHA to avoid chance of conflicts

sarthak-19 committed:
Merge branch 'trunk' into fix/add_cover_annotation_test

westonruter commented:
FYI: I've added the `CODECOV_TOKEN` to the repository secrets.

westonruter reviewed (approved):
I suppose the one job will continue to fail since this PR was opened from a fork. I'll leave it to Felix to approve and merge.

felixarntz reviewed (approved):
Thanks for the PR @sarthak-19, LGTM!

Before we merge though: Doesn't one of us has to configure / set up this project on https://about.codecov.io/ first? 🤔

westonruter commented:
@felixarntz I _believe_ it's all already configured according to the steps listed at https://app.codecov.io/gh/WordPress/performance

The only thing outstanding is the third step, which is to merge into the preferred branch:

> ![image](https://github.com/user-attachments/assets/13fe3a19-6bf9-4ad3-9cd1-5fc291eab53c)

I'll do so now!

westonruter commented:
Alas, no: https://github.com/WordPress/performance/actions/runs/11823769472/job/32943781732

westonruter commented:
I've been chatting with @thelovekesh about what's going on. 

For one, we're missing some additional `--` arg separators since we're passing from `npm` to `composer` and then to `phpunit`. So here:

https://github.com/WordPress/performance/blob/745c145709e8d6eec655c109f472ca77b151178d/.github/workflows/php-test-plugins.yml#L77

https://github.com/WordPress/performance/blob/745c145709e8d6eec655c109f472ca77b151178d/.github/workflows/php-test-plugins.yml#L84

These need to instead be, respectively:

```bash
npm run test-php -- -- --coverage-clover=coverage-${{ github.sha }}.xml 
```

and 

```bash
npm run test-php-multisite -- -- --coverage-clover=coverage-multisite-${{ github.sha }}.xml
```

Additionally, the coverage report for each plugin is getting overwritten to the same XML file. There needs to be separate XML files generated for each plugin.

Lovekesh has some additional thoughts.

westonruter commented:
I've opened https://github.com/WordPress/performance/pull/1652 to disable Codecov for the time being.
